opm/grid/cpgpreprocess/preprocess.c

```diff
@@ -716,29 +717,178 @@ vert_size(const struct grdecl *in,
 
 
 /* ---------------------------------------------------------------------- */
-static int
-is_lefthanded(const struct grdecl *in, const double sign)
+static void
+vertex_coord(const struct grdecl *in,
+             const size_t         pillar,
+             const double         z,
+             double              *coord)
+/* ---------------------------------------------------------------------- */
+{
+    const double *top = &in->coord[6*pillar + 0];
+    const double *bot = &in->coord[6*pillar + 3]; /* == top + 3 */
+
+    const int coincide = fabs(top[2] - bot[2]) < 1.0e-6;
+
+    const double t = coincide
+        ? 0.0 /* coincide => vertical */
+        : (z - top[2]) / (bot[2] - top[2]); /* straight line */
+
+    coord[0] = (1.0 - t)*top[0] + t*bot[0];
+    coord[1] = (1.0 - t)*top[1] + t*bot[1];
+    coord[2] = z;
+}
+
+
+static void
+bounding_box_cross_axes(const double *origin,
+                        const double *i_axis,
+                        const double *j_axis,
+                        double       *cross)
+{
+    cross[0] = (i_axis[1] - origin[1]) * (j_axis[2] - origin[2])
+        -      (i_axis[2] - origin[2]) * (j_axis[1] - origin[1]);
+
+    cross[1] = (i_axis[2] - origin[2]) * (j_axis[0] - origin[0])
+        -      (i_axis[0] - origin[0]) * (j_axis[2] - origin[2]);
+
+    cross[2] = (i_axis[0] - origin[0]) * (j_axis[1] - origin[1])
+        -      (i_axis[1] - origin[1]) * (j_axis[0] - origin[0]);
+}
+
+
+static double
+bounding_box_triple_product(const double *origin,
+                            const double *i_axis,
+                            const double *j_axis,
+                            const double *k_axis)
+{
+    double cross[3];
+
+    bounding_box_cross_axes(origin, i_axis, j_axis, cross);
+
+    return cross[0]*(k_axis[0] - origin[0])
+        +  cross[1]*(k_axis[1] - origin[1])
+        +  cross[2]*(k_axis[2] - origin[2]);
+}
+
+
+/* ---------------------------------------------------------------------- */
+enum CoordinateSystemType { Inconclusive, RightHanded, LeftHanded };
+/* ---------------------------------------------------------------------- */
+
+
+/* ---------------------------------------------------------------------- */
+static enum CoordinateSystemType
+classify_geometry(const double triple)
+/* ---------------------------------------------------------------------- */
+{
+    if (triple > 0.0) {
+        return RightHanded;
+    }
+    else if (triple < 0.0) {
+        return LeftHanded;
+    }
+    else {
+        return Inconclusive;
+    }
+}
+
+
+/* ---------------------------------------------------------------------- */
+static enum CoordinateSystemType
+get_cell_type(const struct grdecl *in,
+              const size_t         i,
+              const size_t         j,
+              const size_t         k,
+              const double         sign,
+              const size_t         off[8])
+/* ---------------------------------------------------------------------- */
```

Comment:
Have been biting my tongue, but now I am lost. Some documentation of these function would really help others.

Reply:
> Some documentation of these function would really help others.

I've pushed an update which adds some Doxygen-like comments to most of the internal functions.